javalib/src/main/scala/java/net/URLDecoder.scala

```diff
@@ -0,0 +1,46 @@
+package java.net
+
+import java.io.UnsupportedEncodingException
+import java.nio.{CharBuffer, ByteBuffer}
+import java.nio.charset.Charset
+
+object URLDecoder {
+  def decode(s: String, enc: String): String = {
+
+    if (!Charset.isSupported(enc)) throw new UnsupportedEncodingException(enc)
+
+    val len = s.length
+    val charsetDecoder = Charset.forName(enc).newDecoder()
+    val charBuffer = CharBuffer.allocate(len)
+
+    var i = 0
+    while (i < len) {
+      s.charAt(i) match {
+        case '+' =>
+          charBuffer.append(' ')
+          i += 1
+
+        case '%' if i + 3 < len =>
```

Comment:
This test makes you handling of incorrect input inconsistent. `"%a"` will return `"%a"`, but `"%at"` will throw. The JavaDoc says we can do either, but we need to be consistent over all types of malformedness.